fdbserver/ptxn/TLogPeekCursor.actor.cpp

```diff
@@ -115,18 +119,116 @@ ACTOR Future<bool> peekRemote(PeekRemoteContext peekRemoteContext) {
+	MergedPeekCursor::CursorHeap* pCursorHeap;
+};
+
+// Add a cursor to the CursorHeap, the cursor must hasRemaining()
+void addCursorToCursorHeap(MergedPeekCursor::CursorContainer::iterator iter, MergedPeekCursor::CursorHeap& heap) {
+	ASSERT((*iter)->hasRemaining());
+	const VersionSubsequenceMutation& mutation = (*iter)->get();
+	heap.emplace(mutation.version, mutation.subsequence, iter);
+}
+
+// Peek remote for multiple cursors, if the cursor is not exhausted, remove it from the incoming cursor list.
+ACTOR Future<bool> peekRemoteForMergedCursor(MergedPeekCursorContext context) {
+	state MergedPeekCursor::CursorContainer& cursorPtrs(*context.pCursorPtrs);
+	state MergedPeekCursor::CursorHeap& cursorHeap(*context.pCursorHeap);
+	state int numCursors(cursorPtrs.size());
+	state std::vector<MergedPeekCursor::CursorContainer::iterator> peekedCursorIter;
+	state std::vector<Future<bool>> peekResult;
+	state int numPeeks(0);
+
+	// Only peek for those locally-exhausted cursors
+	for (std::list<std::unique_ptr<PeekCursorBase>>::iterator iter = std::begin(cursorPtrs);
+	     iter != std::end(cursorPtrs);
+	     ++iter) {
+
+		if (!(*iter)->hasRemaining()) {
+			// Only if the cursor has exhausted its local mutations, it needs to peek data from remote.
+			peekedCursorIter.push_back(iter);
+			peekResult.push_back((*iter)->remoteMoreAvailable());
+			++numPeeks;
+		}
+	}
+
+	// TODO what would be the proper behavior if *ONE* or a few of the cursors failed? I suppose we could fail the
+	// MergedPeekCursor as a whole but is there a better way? e.g. replace it by a new cursor?
+	wait(waitForAll(peekResult));
+
+	for (int i = 0; i < numPeeks; ++i) {
+		if (!peekResult[i].get()) {
+			// The cursor is exhausted, drop from the list
+			cursorPtrs.erase(peekedCursorIter[i]);
+		} else {
+			addCursorToCursorHeap(peekedCursorIter[i], cursorHeap);
+		}
+	}
+
+	// If there are *ANY* remaining cursors in the list, it means they still have remaining mutations to be consumed, so
+	// the return value is *NOT* related with the peek result.
+	return !cursorPtrs.empty();
+}
+
+struct MergedPeekServerTeamCursorContext : public MergedPeekCursorContext {
+	std::unordered_map<StorageTeamID, CursorContainer::iterator>* pMapper;
+};
+
+// In addition to peekRemoteForMergedCursor, update the team ID/cursor mapping
+ACTOR Future<bool> peekRemoteForMergedServerTeamCursor(MergedPeekServerTeamCursorContext context) {
+	bool result = wait(peekRemoteForMergedCursor(static_cast<MergedPeekCursorContext>(context)));
+
+	// Since peekRemoteForMergedCursor will drop exhausted cursors, the team ID - cursor mapping should drop invalid
+	// references to those exhaustd cursors.
+	std::unordered_set<StorageTeamID> inactiveTeamIDs;
+	for (std::unordered_map<StorageTeamID, CursorContainer::iterator>::iterator iter = std::begin(*context.pMapper);
+	     iter != std::end(*context.pMapper);
+	     ++iter) {
+		inactiveTeamIDs.insert(iter->first);
+	}
```

Comment:
Nit: a concise way is
```
for (const auto& [teamID, _] : *context.pMapper) {
    inactiveTeamIDs.insert(teamID);
}
```

Reply:
Actually to my surprise, those without `[]` will not compile but this one would.